File: dazuma/euler/problem_002.ex
defmodule Euler002 do

  def using_loop(max), do: _loop(max, 1, 2, 0)
  defp _loop(max, _prev, cur, sum) when cur > max, do: sum
  defp _loop(max, prev, cur, sum) do
    if rem(cur, 2) == 0, do: sum = sum + cur
    _loop(max, cur, prev + cur, sum)
  end

  def using_streams(max) do
    Stream.iterate({1, 1}, fn {prev, cur} -> {cur, prev + cur} end)
    |> Stream.map(&(elem(&1, 1)))
    |> Stream.filter(&(rem(&1, 2) == 0))
    |> Stream.take_while(&(&1 <= max))
    |> Enum.sum()
  end

end


Euler002.using_streams(4000000) |> IO.puts

File: dazuma/euler/problem_002.rb
module Euler002

  def self.using_loop(max)
    prev, cur, sum = 1,2,0
    until cur > max
      sum += cur if cur % 2 == 0
      prev, cur = cur, prev + cur
    end
    sum
  end


  def self.using_streams(max)
    fibs = Enumerator.new{ |y|
      prev, cur = 1, 1
      loop do
        y << cur
        prev, cur = cur, prev + cur
      end
    }
    fibs.lazy
      .find_all{ |a| a % 2 == 0 }
      .take_while{ |a| a <= max }
      .inject(:+)
  end

end


puts Euler002.using_streams(4000000)
